### DISCUSSION TITLE
Diff check before auto deploys for monorepo

### DISCUSSION THREAD

**Discussion by cyrus-za** [vercel/vercel#5142] [Ideas] [answered]
Following the launch of monorepo support and the hotly discussed topics at #3547 I figured it makes sense to start splitting up some of the points discussed there into their own issues, starting with this feature request: do a diff check on the git integration to see which directories changed and only deploy the services that are affected for these directories.

e.g.
If I have a monorepo with lets say 2x nextjs apps and 1x backend node server, and I push a commit to a branch that has changes to app1 and server1 but does not affect app2, I'd expect only app1 and server1 to be affected.

**Comment by paulogdm** [comment]
Check if https://vercel.com/docs/platform/projects#ignored-build-step can't assist you temporarily.

**Reply by josias-r** [reply]
@paulogdm It seems to me grep is not working. The following two examples always cancel:
`echo master | grep -q master && exit 1 || exit 0`
`echo develop | grep -q master && exit 1 || exit 0`
even though only the second one should, correct?

**Reply by josias-r** [reply]
Nor does something work like ` [[ "master" == "master" ]] && exit 1 || exit 0`. It will throw an error.
`Error: Command failed with ENOENT: [[ "master" == "master" ]] && exit 1 || exit 0`

When rolling out a feature like that, give some info on what the devs can work with :)

**Reply by paulogdm** [reply]
We've just released a guide about this field: https://vercel.com/knowledge/how-do-i-use-the-ignored-build-step-field-on-vercel

**Reply by josias-r** [reply]
@paulogdm Awesome, thanks!

**Reply by osdiab** [reply]
@paulogdm i've tried literally copy and pasting the exact scripts in that article but they never seem to actually work; it would be really helpful if I could see logs from these ignore build step tasks. No idea what's going wrong but the environment variables don't seem to be getting filled in despite clicking the "automatically expose system environment variables" checkbox in the console and running bash scripts that exactly match the examples in the article above.

**Comment by cyrus-za** [comment]
Thanks for this. Is this new? I did not see it when I checked for it a week ago?

I went through it, setup a bash script, but it's not working, despite giving the correct exit code when testing locally and I cant see any logs on the vercel build step. Any help would be appreciated

**Reply by paulogdm** [reply]
It is! The feature is live since we released support for monorepo.

**Comment by cyrus-za** [comment, marked as answer]
@paulogdm I think I figured out why the script is not working. Do you clone with a depth=1 and only the current branch? So no way to compare with master?

**Reply by paulogdm** [reply]
That is correct. We received some requests to customize the command and the team is looking into it.

**Reply by cyrus-za** [reply]
Thanks @paulogdm 

Any update on that? Or any workaround to find out if a certain directory has changed? We got a bunch of services and our builds are really slow with a super long queue at certain times. We've gone and disabled git integration on more than half our services because of this

**Reply by cyrus-za** [reply]
I'd also like to point out that we got severals queues on the same branch.

e.g. if I got branch `A` and branch `B` and I open a PR for both and I push to branch `A` it will queue 5 service deployments (got only 5 turned on for git integration currently). If I then push code to `B` it will queue another 5 service deployments for that branch. If I then push a new commit to `A` before the builds finished, we'll be at 15 queued deployments, and it won't cancel the old queued deployments to `A` despite newer code already queued up.

**Reply by paulogdm** [reply]
You can use `git diff HEAD^ HEAD path/here/` as an example. We are preparing a document with "recipes".

**Reply by cyrus-za** [reply]
@paulogdm that does not work. It seems to be running from the root directory setup in the general settings (which is required due to monorepo git integration), so specifying the path from the root of the repo throws an error with the path shown twice, and doing `git diff HEAD^ HEAD ./` just cancels the build regardless of whether there were actually changes.

**Reply by cyrus-za** [reply]
@paulogdm any update? It's been a few weeks

**Reply by paulogdm** [reply]
Please check https://vercel.com/knowledge/how-do-i-use-the-ignored-build-step-field-on-vercel

**Reply by cyrus-za** [reply]
@paulogdm 
When I copy paste the command from the link you provided, I get the following error (after changing the path to match the relevant service).

Root Directory is set to `/services/asgard`
Ignore Build Step cmd is `git diff HEAD^ HEAD --quiet services/asgard/`
```sh
-- | --
13:35:45.355 | Error: Command failed with exit code 128: git diff HEAD^ HEAD --quiet services/asgard/
13:35:45.355 | fatal: ambiguous argument 'services/asgard/': unknown revision or path not in the working tree.
13:35:45.355 | Use '--' to separate paths from revisions, like this:
13:35:45.355 | 'git <command> [<revision>...] -- [<file>...]'
```

**Reply by cyrus-za** [reply]
UPDATE: I changed ignore cmd to `git diff HEAD^ HEAD --quiet .` and it worked. Seems like it runs the command from within the rot directory of the service, not the root directory of the repo

**Reply by cyrus-za** [reply]
@paulogdm I now have a new problem. We are running into a rate limit. It appears that each cancelled build for each of our projects in the monorepo count towards the rate limit. Can you please fix this?

**Reply by paulogdm** [reply]
For this type of request, I would recommend getting in touch with us: https://vercel.com/support/request

The rate-limit exists for several reasons, including capacity planning, risk mitigation, misusage prevention, etc.

**Reply by cyrus-za** [reply]
Yes, but in this case it is none of those reasons. Its the fact that each and every canceled build counts towards usage, which is obviously a bug. And I did reach out to support. No response yet

**Reply by cyrus-za** [reply]
@paulogdm new issue. The git command you recommend to check the head only checks the last commit, not the entire PR which is not what we want and due to depth=1 we cant check anything else. This has resulted in changes not deployed, not only in PR previews but even on master and production branches. We we wrote a bash script to at least always deploy on master and production (not ideal as not all services would have changed when we do a merge). 

The problem is that bash script is not working. Even if we do a exit 0 it will still build. 

script below. Please let me know if I am doing something wrong. I debugged locally and it gives the right exit code. No way to debug it on vercel as it does not log the output, or even logs from within the script.

```sh
#!/bin/bash
#
# Print commands as they are executed (without varible expansion because secrets)
set +v
set -e

DEFAULT="\\033[39m"
error() { echo -e "\\033[31m$1$DEFAULT"; }
success() { echo -e "\\033[32m$1$DEFAULT"; }
info() { echo -e "\\033[34m$1$DEFAULT"; }

deployToVercel() {
  changed=`git diff HEAD^ HEAD $PWD`
  branch=$(git branch --show-current)

  if [ "$changed" == "diff" ] || [ "$branch" == "master" ] || [ "$branch" == "production" ]
  then
    exit 1
  fi
    success "*** No changes since the last merge, so not deploying ***"
  exit 0
}

deployToVercel $*
```

**Reply by homerjam** [reply]
> UPDATE: I changed ignore cmd to `git diff HEAD^ HEAD --quiet .` and it worked. Seems like it runs the command from within the rot directory of the service, not the root directory of the repo

The placeholder in the dashboard is confusing - I was under the impression that this was being used unless I defined a command myself. It'd be less confusing to show the example above the input, or flag it very clearly at the top of the docs.

**Reply by TheRusskiy** [reply]
@cyrus-za 
Did you have any luck with the script? Thank you

**Reply by paulogdm** [reply]
Related: https://github.com/vercel/vercel/discussions/5171#discussioncomment-139791

**Reply by cyrus-za** [reply]
@TheRusskiy nope. we just manually triggering dereploys if we got multiple services having changed in different commits. Its a bit of a joke. Paying for a enterprise package but nobody at support team can help us

**Reply by paulogdm** [reply]
Hey @cyrus-za . This is very unexpected from us. Can you give me the support tickets? Feel free to email me at paulo@vercel.com as well.

**Reply by paulogdm** [reply]
@TheRusskiy I just did an update to print the variables for better debugging, for example: https://vercel.com/knowledge/how-do-i-use-the-ignored-build-step-field-on-vercel

**Reply by ironhee** [reply]
> @paulogdm I now have a new problem. We are running into a rate limit. It appears that each cancelled build for each of our projects in the monorepo count towards the rate limit. Can you please fix this?

I'm also having a rate limit problem. Do you have any solution to this issue? @paulogdm

**Reply by ironhee** [reply]
The Ignored Build Step works fine. The only problem is that ignored deployments are included in the rate limit.

**Reply by paulogdm** [reply]
We have an internal task to remove the rate-limit for canceled builds.

**Reply by cyrus-za** [reply]
@paulogdm can we also get it to not show cancelled builts in the list, or on github integrations? Its a little annoying to see all our apps when only a single app was affected

**Reply by paulogdm** [reply]
Absolutely @cyrus-za . The current way is just sub-optimal DX-wise. Please talk with your CSM (Evelynn) about this so we can make sure this feature gets released.
